velox/type/HugeInt.cpp

```diff
@@ -16,6 +16,32 @@
 
 #include "velox/type/HugeInt.h"
 
+namespace facebook::velox {
+
+int128_t HugeInt::parse(const std::string& str) {
+  int128_t result = 0;
+  bool negative = false;
+  size_t idx = 0;
+
+  VELOX_CHECK(!str.empty(), "Empty string cannot be converted to int128_t.");
+
+  if (str[0] == '-') {
+    negative = true;
+    idx = 1;
+  }
+
+  for (; idx < str.size(); ++idx) {
+    VELOX_CHECK(
+        str[idx] >= '0' && str[idx] <= '9',
+        "Invalid character {} in the string.",
+        str[idx])
+    result = result * 10 + (str[idx] - '0');
```

Comment:
This is failing under ASAN when parsing the mininum value: "signed integer overflow: 0x7ffffffffffffffffffffffffffffff8 + 8 cannot be represented in type '__int128'".

Reply:
Fixed, by adding early return logic in HugeInt::parse to avoid the potential *overflow* caused by *Type Widening* of int128_t making it more robust.
```C++
int128_t max = std::numeric_limits<int128_t>::max();
int128_t min = std::numeric_limits<int128_t>::min();
for (; idx < str.size(); ++idx) {
   ...
    int cur = str[idx] - '0';
    if ((result > max / 10) || (result == (max / 10) && cur > (max % 10))) {
      return negative ? min : max;
    }
    result = result * 10 + cur;
}
return negative ? -result : result;
```